### PR TITLE
fix: Correctly decode optional struct fields

### PR DESCRIPTION
**Pull request by frederikrothenberger**
# Description

Fixes incorrect decoding of optional struct fields.

# How Has This Been Tested?

A unit test for this corner case has been added.

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

File: packages/candid/src/idl.ts [hunk 1]
```diff
@@ -924,15 +924,29 @@ export class RecordClass extends ConstructType<Record<string, any>> {
...
+        (expectType instanceof OptClass || expectType instanceof ReservedClass) &&
+        idx < this._fields.length &&
+        idlLabelToId(this._fields[idx][0]) !== idlLabelToId(hash)
+      ) {
+        x[expectKey] = [];
+        idx++;
+        [expectKey, expectType] = this._fields[idx];
+      }
+
       if (idx >= this._fields.length || idlLabelToId(this._fields[idx][0]) !== idlLabelToId(hash)) {
```
[the remark this answers is not in the document]
Comment (frederikrothenberger) [reply packages/candid/src/idl.ts]: Unfortunately not because then we would go to the next pair of `const [hash, type] of record._fields` without having processed the current one. I have now switched the whole thing to a `while` loop.

**Review by krpeacock — APPROVED** [review]
